File: services/api-go/internal/features/transforms.go
package features

import (
	"math"
	"regexp"
	"strings"
	"unicode"
)

const (
	GenderNone   = 0.0
	GenderMen    = 1.0
	GenderWomen  = 2.0
	GenderBoys   = 3.0
	GenderGirls  = 4.0
	GenderUnisex = 5.0
)

// Tokenize splits on Unicode word boundaries (matches Python re.findall(r"\w+")
// when both sides interpret \w as Unicode word). All tokens are lowercased.
func Tokenize(s string) []string {
	if s == "" {
		return nil
	}
	out := make([]string, 0, 8)
	var sb strings.Builder
	for _, r := range s {
		if unicode.IsLetter(r) || unicode.IsDigit(r) || r == '_' {
			sb.WriteRune(unicode.ToLower(r))
		} else {
			if sb.Len() > 0 {
				out = append(out, sb.String())
				sb.Reset()
			}
		}
	}
	if sb.Len() > 0 {
		out = append(out, sb.String())
	}
	return out
}

func QueryLengthTokens(query string) float64 {
	return float64(len(Tokenize(query)))
}

func QueryHasAny(query string, vocab map[string]struct{}) float64 {
	if len(vocab) == 0 || query == "" {
		return 0
	}
	for _, t := range Tokenize(query) {
		if _, ok := vocab[t]; ok {
			return 1
		}
	}
	return 0
}

// Identical regex to the Python reference impl (transforms.py).
var sizeRe = regexp.MustCompile(`(?i)\b\d+(?:\.\d+)?\s?(?:oz|ml|gb|tb|in|cm|mm|kg|lb|l|g)\b`)

func QueryHasSizePattern(query string) float64 {
	if query == "" {
		return 0
	}
	if sizeRe.FindStringIndex(query) != nil {
		return 1
	}
	return 0
}

func QueryAffordabilityIntent(query string) float64 {
	tokens := Tokenize(query)
	for i, t := range tokens {
		if affordabilityToken(t) {
			return 1
		}
		if i == 0 {
			continue
		}
		prev := tokens[i-1]
		if (prev == "low" || prev == "lower" || prev == "lowest") && (t == "cost" || t == "price") {
			return 1
		}
	}
	return 0
}

func AffordabilityPriceScore(queryAffordability float64, priceCents int) float64 {
	if queryAffordability == 0 || priceCents <= 0 {
		return 0
	}
	return 1 / math.Log1p(float64(priceCents))
}

func QueryGenderIntent(query string) float64 {
	found := GenderNone
	for _, t := range Tokenize(query) {
		g := queryGenderToken(t)
		if g == GenderNone {
			continue
		}
		if found != GenderNone && found != g {
			return GenderNone
		}
		found = g
	}
	return found
}

func affordabilityToken(t string) bool {
	switch t {
	case "affordable", "affordability", "cheap", "cheaper", "cheapest",
		"budget", "inexpensive", "economical", "value", "price", "priced",
		"pricing", "cost", "costs":
		return true
	default:
		return false
	}
}

func ProductGender(categoryPath []string, title string) float64 {
	for _, part := range categoryPath {
		if gender := ProductGenderLabel(part); gender != GenderNone {
			return gender
		}
	}
	return titleGender(title)
}

func ProductGenderLabel(label string) float64 {
	switch strings.ToLower(strings.TrimSpace(label)) {
	case "men":
		return GenderMen
	case "women":
		return GenderWomen
	case "boys":
		return GenderBoys
	case "girls":
		return GenderGirls
	case "unisex":
		return GenderUnisex
	default:
		return GenderNone
	}
}

func GenderIntentMatch(queryGender, productGender float64) float64 {
	if queryGender != GenderNone && queryGender == productGender {
		return 1
	}
	if (queryGender == GenderMen || queryGender == GenderWomen) && productGender == GenderUnisex {
		return 0.5
	}
	return 0
}

func GenderIntentMismatch(queryGender, productGender float64) float64 {
	if productGender == GenderUnisex {
		return 0
	}
	if queryGender != GenderNone && productGender != GenderNone && queryGender != productGender {
		return 1
	}
	return 0
}

func ProductBrandMatch(query, brand string) float64 {
	queryTokens := tokenSet(BrandTokens(query))
	brandTokens := tokenSet(BrandTokens(brand))
	for t := range brandTokens {
		if _, ok := queryTokens[t]; ok {
			return 1
		}
	}
	return 0
}

func ProductBrandTokenOverlap(query, brand string) float64 {
	queryTokens := tokenSet(BrandTokens(query))
	brandTokens := tokenSet(BrandTokens(brand))
	if len(brandTokens) == 0 {
		return 0
	}
	matched := 0
	for t := range brandTokens {
		if _, ok := queryTokens[t]; ok {
			matched++
		}
	}
	return float64(matched) / float64(len(brandTokens))
}

func SubbrandTitleMatch(query, title string) float64 {
	queryTokens := tokenSet(BrandTokens(query))
	if len(queryTokens) == 0 {
		return 0
	}
	titleTokens := Tokenize(title)
	for subbrand := range queryTokens {
		if !knownSubbrand(subbrand) {
			continue
		}
		if titleContainsSubbrandAlias(titleTokens, subbrand) {
			return 1
		}
	}
	return 0
}

func BrandFamilyMatch(query, brand, title string) float64 {
	queryTokens := tokenSet(BrandTokens(query))
	brandTokens := tokenSet(BrandTokens(brand))
	if len(queryTokens) == 0 || len(brandTokens) == 0 {
		return 0
	}
	for t := range brandTokens {
		if _, ok := queryTokens[t]; ok {
			return 1
		}
	}
	for subbrand := range queryTokens {
		if !subbrandParentBrandMatch(subbrand, brandTokens) {
			continue
		}
		if SubbrandTitleMatch(query, title) == 1 {
			return 1
		}
	}
	return 0
}

func ProductColorMatch(query, color string) float64 {
	return ProductBrandMatch(query, color)
}

func BrandTokens(text string) []string {
	tokens := Tokenize(text)
	out := make([]string, 0, len(tokens))
	for _, token := range tokens {
		if brandStopToken(token) {
			continue
		}
		out = append(out, token)
	}
	return out
}

func brandStopToken(token string) bool {
	switch token {
	case "accessories", "accessory", "active", "athletic", "basketball",
		"boy", "boys", "clothing", "fashion", "girl", "girls", "jewelry",
		"men", "mens", "running", "shoe", "shoes", "sneaker", "sneakers",
		"sports", "team", "watch", "watches", "woman", "women", "womens":
		return true
	case "unisex":
		return true
	default:
		return false
	}
}

func QueryTokenCoverage(query, text string, ignored map[string]struct{}) float64 {
	queryTokens := Tokenize(query)
	if len(queryTokens) == 0 {
		return 0
	}
	textTokens := tokenSet(Tokenize(text))
	keptTokens := make([]string, 0, len(queryTokens))
	for _, t := range queryTokens {
		if _, skip := ignored[t]; skip {
			continue
		}
		keptTokens = append(keptTokens, t)
	}
	if len(keptTokens) == 0 {
		keptTokens = queryTokens
	}
	matched := 0
	for _, t := range keptTokens {
		if _, ok := textTokens[t]; ok {
			matched++
		}
	}
	return float64(matched) / float64(len(keptTokens))
}

func ExactQueryPhraseMatch(query, text string) float64 {
	q := strings.Join(Tokenize(query), " ")
	haystack := strings.Join(Tokenize(text), " ")
	if q != "" && strings.Contains(haystack, q) {
		return 1
	}
	return 0
}

func TokenOverlapFraction(query, text string) float64 {
	queryTokens := tokenSet(Tokenize(query))
	textTokens := tokenSet(Tokenize(text))
	if len(textTokens) == 0 {
		return 0
	}
	matched := 0
	for t := range textTokens {
		if _, ok := queryTokens[t]; ok {
			matched++
		}
	}
	return float64(matched) / float64(len(textTokens))
}

func tokenSet(tokens []string) map[string]struct{} {
	out := make(map[string]struct{}, len(tokens))
	for _, t := range tokens {
		out[t] = struct{}{}
	}
	return out
}

func knownSubbrand(subbrand string) bool {
	switch subbrand {
	case "jordan":
		return true
	default:
		return false
	}
}

func subbrandParentBrandMatch(subbrand string, brandTokens map[string]struct{}) bool {
	switch subbrand {
	case "jordan":
		_, ok := brandTokens["nike"]
		return ok
	default:
		return false
	}
}

func titleContainsSubbrandAlias(titleTokens []string, subbrand string) bool {
	switch subbrand {
	case "jordan":
		return containsTokenSequence(titleTokens, []string{"air", "jordan"}) ||
			containsTokenSequence(titleTokens, []string{"jordan"})
	default:
		return false
	}
}

func containsTokenSequence(tokens, sequence []string) bool {
	if len(sequence) == 0 || len(sequence) > len(tokens) {
		return false
	}
	for i := 0; i <= len(tokens)-len(sequence); i++ {
		match := true
		for j, want := range sequence {
			if tokens[i+j] != want {
				match = false
				break
			}
		}
		if match {
			return true
		}
	}
	return false
}

func queryGenderToken(t string) float64 {
	switch t {
	case "men", "mens", "man", "male":
		return GenderMen
	case "women", "womens", "woman", "female":
		return GenderWomen
	case "boys", "boy":
		return GenderBoys
	case "girls", "girl":
		return GenderGirls
	case "unisex":
		return GenderUnisex
	default:
		return GenderNone
	}
}

func titleGender(title string) float64 {
	found := GenderNone
	for _, t := range Tokenize(title) {
		g := queryGenderToken(t)
		if g == GenderNone {
			continue
		}
		if g == GenderUnisex {
			return GenderUnisex
		}
		if found != GenderNone && found != g {
			if (found == GenderMen && g == GenderWomen) || (found == GenderWomen && g == GenderMen) {
				return GenderUnisex
			}
			return GenderNone
		}
		found = g
	}
	return found
}

File: pipelines/pipelines/features/transforms.py
"""Reference Python implementations for INTERACTION features.

Specced in libs/schema/transforms.md. Phase 5 mirrors these in Go and a
parity test runs both impls on the same fixtures.
"""

from __future__ import annotations

import re
import math

# Tokenize on \w+ (Unicode word characters), lowercase. Matches Go regexp default.
_TOKEN_RE = re.compile(r"\w+", re.UNICODE)

# Size-with-unit pattern: 12oz, 1.5L, 16gb, 7in, 9.5 mm, etc.
_SIZE_RE = re.compile(
    r"\b\d+(?:\.\d+)?\s?(?:oz|ml|gb|tb|in|cm|mm|kg|lb|l|g)\b",
    re.IGNORECASE,
)

GENDER_NONE = 0.0
GENDER_MEN = 1.0
GENDER_WOMEN = 2.0
GENDER_BOYS = 3.0
GENDER_GIRLS = 4.0

_QUERY_GENDER_TOKENS = {
    "men": GENDER_MEN,
    "mens": GENDER_MEN,
    "man": GENDER_MEN,
    "male": GENDER_MEN,
    "women": GENDER_WOMEN,
    "womens": GENDER_WOMEN,
    "woman": GENDER_WOMEN,
    "female": GENDER_WOMEN,
    "boys": GENDER_BOYS,
    "boy": GENDER_BOYS,
    "girls": GENDER_GIRLS,
    "girl": GENDER_GIRLS,
}

_CATEGORY_GENDER_VALUES = {
    "men": GENDER_MEN,
    "women": GENDER_WOMEN,
    "boys": GENDER_BOYS,
    "girls": GENDER_GIRLS,
}


def tokenize(s: str | None) -> list[str]:
    if not s:
        return []
    return [t.lower() for t in _TOKEN_RE.findall(s)]


def query_length_tokens(query: str | None) -> float:
    return float(len(tokenize(query)))


def query_has_brand(query: str | None, known_brands_lower: frozenset[str]) -> float:
    return 1.0 if (set(tokenize(query)) & known_brands_lower) else 0.0


def query_has_color(query: str | None, known_colors_lower: frozenset[str]) -> float:
    return 1.0 if (set(tokenize(query)) & known_colors_lower) else 0.0


def query_has_category_token(query: str | None, known_category_tokens: frozenset[str]) -> float:
    return 1.0 if (set(tokenize(query)) & known_category_tokens) else 0.0


def query_has_size_pattern(query: str | None) -> float:
    return 1.0 if (query and _SIZE_RE.search(query)) else 0.0


_AFFORDABILITY_TOKENS = {
    "affordable",
    "affordability",
    "cheap",
    "cheaper",
    "cheapest",
    "budget",
    "inexpensive",
    "economical",
    "value",
    "price",
    "priced",
    "pricing",
    "cost",
    "costs",
}
_AFFORDABILITY_PREFIXES = {"low", "lower", "lowest"}
_AFFORDABILITY_NOUNS = {"cost", "price"}


def query_affordability_intent(query: str | None) -> float:
    tokens = tokenize(query)
    for i, token in enumerate(tokens):
        if token in _AFFORDABILITY_TOKENS:
            return 1.0
        if i and tokens[i - 1] in _AFFORDABILITY_PREFIXES and token in _AFFORDABILITY_NOUNS:
            return 1.0
    return 0.0


def affordability_price_score(query_affordability: float, price_cents: int | float | None) -> float:
    price = float(price_cents or 0)
    if not query_affordability or price <= 0:
        return 0.0
    return float(1 / math.log1p(price))


def query_gender_intent(query: str | None) -> float:
    found = {g for t in tokenize(query) if (g := _QUERY_GENDER_TOKENS.get(t))}
    return found.pop() if len(found) == 1 else GENDER_NONE


def product_gender(category_path: list[str] | tuple[str, ...] | None) -> float:
    if not category_path:
        return GENDER_NONE
    for part in category_path:
        gender = _CATEGORY_GENDER_VALUES.get(str(part).strip().lower())
        if gender:
            return gender
    return GENDER_NONE


def gender_intent_match(query_gender: float, product_gender_value: float) -> float:
    return 1.0 if query_gender and query_gender == product_gender_value else 0.0


def gender_intent_mismatch(query_gender: float, product_gender_value: float) -> float:
    return 1.0 if query_gender and product_gender_value and query_gender != product_gender_value else 0.0


def product_brand_match(query: str | None, brand: str | None) -> float:
    query_tokens = set(tokenize(query))
    brand_tokens = set(tokenize(brand))
    return 1.0 if query_tokens and brand_tokens and bool(query_tokens & brand_tokens) else 0.0


def product_brand_token_overlap(query: str | None, brand: str | None) -> float:
    query_tokens = set(tokenize(query))
    brand_tokens = set(tokenize(brand))
    if not brand_tokens:
        return 0.0
    return float(len(query_tokens & brand_tokens) / len(brand_tokens))


def product_color_match(query: str | None, color: str | None) -> float:
    return product_brand_match(query, color)


def query_token_coverage(
    query: str | None,
    text: str | None,
    ignored_tokens: frozenset[str] = frozenset(),
) -> float:
    query_tokens = tokenize(query)
    q = [t for t in query_tokens if t not in ignored_tokens]
    if not q:
        q = query_tokens
    if not q:
        return 0.0
    text_tokens = set(tokenize(text))
    return float(sum(1 for t in q if t in text_tokens) / len(q))


def exact_query_phrase_match(query: str | None, text: str | None) -> float:
    q = " ".join(tokenize(query))
    haystack = " ".join(tokenize(text))
    return 1.0 if q and q in haystack else 0.0


def token_overlap_fraction(query: str | None, text: str | None) -> float:
    query_tokens = set(tokenize(query))
    text_tokens = set(tokenize(text))
    if not text_tokens:
        return 0.0
    return float(len(query_tokens & text_tokens) / len(text_tokens))


def product_category_token_overlap(query: str | None, category_text: str | None) -> float:
    return token_overlap_fraction(query, category_text)
